db/hints/manager.hh

```diff
@@ -135,6 +136,7 @@ public:
 
         private:
             std::list<sstring> _segments_to_replay;
+            std::unordered_set<sstring> _segments_to_replay_set;
```

Comment:
This is going to at least double the memory consumption occupied by segments file names and most of the time we don't need this hence this is going to be memory wasted.

Let's remember what you really what you want to achieve here the way I understand this: you want to be able to send out all hints *aggregated so far*.  While you are going to be sending those hints new hints should be allowed to be generated but you don't want them to be taken into an account.

So, why don't we do the operation that this patch implements only when we actually need it? Namely when we need to sync:
 1) We force a commitlog object closing.
 2) Push all segments that we have so far into _segments_to_replay.
 3) Resolve the `future` when all segments from (2) above are sent.

This way we don't need to permanently maintain _segments_to_replay_set and will only need it during (2).

Reply:
@avikivity @haaawk The above is a blocking issue too.
@piodul FYI

Reply:
Yes. As far as I know, @piodul is planning to fix that. Probably it would be better to answer sooner to let you know.

Reply:
@vladzcloudius Done.

I wonder if we could avoid triggering a flush, though. In current implementation, when `wait_until_hints_are_replayed` is called, all hint queues will be flushed at the same time. Could this cause problem? Maybe we should limit concurrent flushes somehow?

Reply:
Thanks, @piodul. I'll take a look.

> I wonder if we could avoid triggering a flush, though. 

If you want to be on a formal side of things I think you have to make it - you want to make sure there are no "old" hints stuck in the memory (commitlog) since the corresponding mutations may be reducing the amount of repair work.

However, let's stop being mathematicians for a second and imagine that we are practical dudes like @gleb-cloudius. Considering that the amount of mutations that may possibly sit in memory is limited by the size of a single (AFAIR) commitlog segment (32MB) this is exactly what caps the amount of "damage" the not-flushing may cause you (multiple the number of per-destination queues multiple the number of shards). But practically, most of those queues are going to be empty.

So, from the practical perspective this flush doesn't change much and IMO can be safely skipped. So, all you'd need to do is to pick up names of all already existing segments.

However, if you decide to go that path keep in mind that unless you shut down the commitlog object there is going to be at least one segment that is not yet sealed and you want to exclude it from the bunch. 

Unless APIs required for the above exist it would require patching commitlog (which we usually prefer to avoid). 

In such a case you may think about it from the opposite angle - how bad is that flushing going to be? See more below.

> Could this cause problem? Maybe we should limit concurrent flushes somehow?

Theoretically - yes, it may cause an issue. If we have big nodes (with many shards) and some not trivial amount of nodes then there may be a rather significant amount of such queues. So, yes, you may indeed want to limit the concurrency of those flushes in order to avoid an I/O spike. I think that keeping a per-shard max concurrency as 1 (flushing them serially) is going to be just fine - you can do a quick test and see how long it takes to flush, say 100 commitlog segments, one after another - and the total amount of queues on each shards is limited by the total number of nodes in the cluster minus 1. So, if you flush 100 files in less than 5 seconds - I think you are golden. 
Why serially? - Because it's the least impactful on the I/O system and if it's fast enough no reason to go higher. And this event (repair) is a very rare and a VERY LONG event - so, what if takes a few more seconds to ignite as long as we keep things simple and clean... ;)

Reply:
> However, if you decide to go that path keep in mind that unless you shut down the commitlog object there is going to be at least one segment that is not yet sealed and you want to exclude it from the bunch.
>
> Unless APIs required for the above exist it would require patching commitlog (which we usually prefer to avoid).

I agree that it's not required to be strict. The first version of this PR was like that, but it could skip more than one segment - it skipped segments created by the current commitlog instance. So, it could skip hints created during at most 10s period.

I thought about the solution you proposed where we look at the currently existing segments, and I'd rather not do it. The list of the currently existing segments will usually contain some segments which are still managed by currently existing commitlog instance. Conceptually, the commitlog might be still doing something with those segments, so I'm worried that something could break in the future.

> In such a case you may think about it from the opposite angle - how bad is that flushing going to be? See more below.

> I think that keeping a per-shard max concurrency as 1 (flushing them serially) is going to be just fine - you can do a quick test and see how long it takes to flush, say 100 commitlog segments, one after another - and the total amount of queues on each shards is limited by the total number of nodes in the cluster minus 1. So, if you flush 100 files in less than 5 seconds - I think you are golden.

I changed the implementation so that on each shard, we flush at most one hint queue at a time when creating a hint sync point.

To see how long it will take, I did some tests (locally on my machine) using the following scenario:

- Created a large cluster,
- Created a keyspace with RF equal to the cluster size,
- Shut down all nodes apart from one or two,
- Started a write workload so that hint queues are created and hints are written to them,
- Ran `nodetool repair` in order to trigger hints flushing towards each other node.

Results look rather good:

- 12 nodes, 1 shard per node: 50-500ms
- 12 nodes, 6 shards per node: 1-2s
- 90 nodes, 1 shard per node: 1-2s